File: go/198.go
/*
DP, dp[i] = max(dp[i-1], dp[i-2] + nums[i])
*/
func rob(nums []int) int {
    dp := make([]int, len(nums)+2)
    for i := 2; i < len(dp); i++ {
        dp[i] = max(dp[i-1], dp[i-2] + nums[i-2])
    }
    return dp[len(dp)-1]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: python/198.py
# DP problem. f(n) = max(f(n-1), f(n-2)+c_n).
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dp = [0] * (len(nums) + 2)
        dp[0] = dp[1] = 0
        for i in range(len(nums)):
            dp[i+2] = max(dp[i] + nums[i], dp[i+1])
        return dp[len(nums) + 1]